### ISSUE TITLE
QgsPostgresConn::retrieveLayerTypes scans full table by default

### ISSUE THREAD

**Issue by pramsey** [qgis/QGIS#58742]
### Feature description

https://github.com/qgis/QGIS/blob/e7c045529664cd3e00225dd59bd0b16d10d44acd/src/providers/postgres/qgspostgresconn.cpp#L2152

A full scan of the tables is the default action when connecting to a PostGIS database, and when a database has some very large tables, or FDW tables (slow to query sometimes) that can result in a hung QGIS. Probably "full scan" is the wrong default, and maybe never the right answer. 

* At a first go, just having "Use estimated table metadata" checked instead of unchecked.
* Instead of taking "first 100 rows", using the PostgreSQL TABLESAMPLE facility can get a more representative sample at a lower cost, which might make having "default is sample" more palatable in terms of outputs.

**Comment by bertday** [comment]
I came to report the same behavior and stumbled on this issue. When I connect to a Postgres DB containing tables with lots of rows, it takes quite a while to list the tables in the QGIS browser.

I took a look in `pg_stat_activity` and saw QGIS is making a query like this for every table in the database:

```sql
SELECT 0, array_agg(DISTINCT st_srid("geom"::geometry)::text || ':' || UPPER(geometrytype("geom"::geometry))  || ':' || ST_Zmflag("geom"::geometry))  FROM (SELECT "geom" FROM "public"."a_big_table_with_lots_of_rows") AS _unused;
```

I believe that's the same full scan @pramsey described above, but the underlying SQL. I can see how it makes sense for a complete understanding of the data, but would agree with Paul that the overhead may not be justified in most cases.

**Comment by gangerang** [comment]
Yes, I experience this issue! Can take 5 mins or so of waiting for these scans to occur when first opening up my connected PostGIS db. I had assumed it was because I hadn't configured my tables correctly to explicitly specify what type of geometry my columns were.

**Comment by WCMC-vblanque** [comment]
Hi mates, 

Thanks @pramsey to suggest **"Use estimated table metadata"** but it is not speeding up enough in my case.

I have been facing similar lag time... I have many schemas, many tables with many rows... and it is becoming a bottleneck use QGIS for visual checks of my queries. 

- Do you know if an improvement is already in the pipeline so that I could test it?  
- Is there a patch I could use meanwhile?
- Do you have a PyQGIS script to load specific table?  I have tried to develop a quick one, but did not succeed until now. If you have one to share, it would be very helpful.
- Do you remind if any older version of QGIS was dealing better with big database?  I tried with 2.18 but it seems the authentification system has changed and I cannot currently connect QGIS 2.18 with my current PG16.

Comments:
_As a workaround I might create a new user with access to specific schema and tables. Let's see how it goes... and let you know._
_It looks like PostGIS toolbox is a bit more performant in listing schemas and related tables and views_
_QGIS is an amazing open source software I am using since 2006. I am thinking to donate regularly to support the great community._

Last but not least I would like to help better how to debug, which debugger, log or analytics, I could share here... 

best regards, 

Vincent

**Comment by WCMC-vblanque** [comment]
> o of waiting for these scans to occur when first opening up my connected PostGIS db. I had assumed it was because I hadn't configured my tables correctly to explicitly specify what type of geometry my columns were.

@gangerang 

Here it could take up to 30 minutes even if I use _wildcard_ or _regexp_ to filter upfront the list of schema, tables or views.

Is there a way to insert timestamp (start, end) so that I can track the elapsed time on this... Any idea where to insert this bit? 

![Image](https://github.com/user-attachments/assets/53e6f5ff-874e-41d7-82f7-bd87ecb64257)

**Comment by WCMC-vblanque** [comment]
To share more inputs after checking  `pg_stat_activity`, these many similar queries are **heavy lagging** the connection between QGIS and postgres.

`SELECT cla.relname, nsp.nspname, cla.relkind, pg_get_userbyid(relowner), cla.reltuples, cla.relpages, pg_catalog.obj_description(cla.oid), CASE WHEN geo.f_geometry_column IS NOT NULL THEN geo.f_geometry_column ELSE att.attname END, CASE WHEN geo.type IS NOT NULL THEN geo.type ELSE textin(regtypeout(att.atttypid::regtype)) END, geo.coord_dimension, geo.srid FROM pg_class AS cla JOIN pg_namespace AS nsp ON nsp.oid = cla.relnamespace JOIN pg_attribute AS att ON att.attrelid = cla.oid AND ... `
